File: Company 1 : Microsoft/Day 2/5. Minimum Moves to Equal Array Elements/code.cpp
class Solution {
public:
    int minMoves2(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int ans = 0;
        for(int i=0; i<nums.size(); i++)
            if(nums[i] != nums[nums.size()/2])
                ans += (nums[nums.size()/2]>nums[i])?(nums[nums.size()/2] - nums[i]):(nums[i]-nums[nums.size()/2]);
        return ans;
    }
};

File: Company 1 : Microsoft/Day 2/5. Minimum Moves to Equal Array Elements/code.java
class Solution {
    public int minMoves2(int[] nums) {
        int length = nums.length, result = 0, median = helper(nums, 0, length - 1, (length - 1) / 2);
        for (int x : nums) {
            result += Math.abs(x - median);
        }
        return result;
    }
    private int helper(int[] q, int l, int r, int k) {
        if (l >= r) 
            return q[r];
        int x = q[l + r >> 1], i = l - 1, j = r + 1;
        while (i < j) {
            while (q[++i] < x) ;
            while (q[--j] > x) ;
            if (i < j) {
                int t = q[i];
                q[i] = q[j];
                q[j] = t;
            }
        }
        if (k <= j) 
            return helper(q, l, j, k);
        return helper(q, j + 1, r, k);
    }
}
